### ISSUE TITLE
Java NodeEvaluator2 GC error?

### ISSUE THREAD

**Issue by crofty** [google/or-tools#885]
Hi,
I'm seeing crashes in my program which uses the Java wrapper.

After reading around I suspect that it may be due to a garbage collector issue with NodeEvaluator2.  In fact this issue may be a duplicate of: https://github.com/google/or-tools/issues/473

I can reproduce the issue every time by modifying the `CapacitatedVehicleRoutingProblemWithTimeWindows.java` example and adding the following inside the manhattanCallback, line 180:
`System.out.println("test");`

Adding this line causes the example to fail with the following error:

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x000000012d5b046b, pid=93072, tid=9731
#
# JRE version: Java(TM) SE Runtime Environment (10.0.2+13) (build 10.0.2+13)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (10.0.2+13, mixed mode, tiered, compressed oops, g1 gc, bsd-amd64)
# Problematic frame:
# C  [libjniortools.jnilib+0x4e46b]  _MemberResultCallback_0_2<false, long long, NodeEvaluator2, IntType<operations_research::RoutingNodeIndex_tag_, int>, IntType<operations_research::RoutingNodeIndex_tag_, int>, void>::Run(IntType<operations_research::RoutingNodeIndex_tag_, int>, IntType<operations_research::RoutingNodeIndex_tag_, int>)+0x1b
#
```

**Comment by intelLigenJ** [comment]
I have the same problem!

**Comment by crofty** [comment]
@Mizux I tried with v6.10 and have been able to get that to crash consistently too.  To do so, I added the following line at line 211 of `examples/java/CapacitatedVehicleRoutingProblemWithTimeWindows.java`

```.java
System.out.println("test");
```
Adding this line and running `make run SOURCE=examples/java/CapacitatedVehicleRoutingProblemWithTimeWindows.java`

will produce an error of:

```
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x0000000137e4919b, pid=5542, tid=6403
#
# JRE version: Java(TM) SE Runtime Environment (10.0.2+13) (build 10.0.2+13)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (10.0.2+13, mixed mode, tiered, compressed oops, g1 gc, bsd-amd64)
# Problematic frame:
# C  [libjniortools.jnilib+0x5319b]  _MemberResultCallback_0_2<false, long long, NodeEvaluator2, IntType<operations_research::RoutingNodeIndex_tag_, int>, IntType<operations_research::RoutingNodeIndex_tag_, int>, void>::Run(IntType<operations_research::RoutingNodeIndex_tag_, int>, IntType<operations_research::RoutingNodeIndex_tag_, int>)+0x1b
#
# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again

```

The modified section of code now reads:

```.java
    // Setting up vehicles
    for (int vehicle = 0; vehicle < numberOfVehicles; ++vehicle) {
      final int costCoefficient = vehicleCostCoefficients.get(vehicle);
      NodeEvaluator2 manhattanCostCallback = new NodeEvaluator2() {
        @Override
        public long run(int firstIndex, int secondIndex) {
          try {
            Pair<Integer, Integer> firstLocation = locations.get(firstIndex);
            Pair<Integer, Integer> secondLocation = locations.get(secondIndex);
            System.out.println("test"); // ADDED THIS!
            return costCoefficient
                * (Math.abs(firstLocation.first - secondLocation.first)
                      + Math.abs(firstLocation.second - secondLocation.second));
          } catch (Throwable throwed) {
            logger.warning(throwed.getMessage());
            return 0;
          }
        }
      };
      model.setArcCostEvaluatorOfVehicle(manhattanCostCallback, vehicle);

```

**Comment by lperron** [comment]
You can keep a reference to the callbacks outside the loop.
I guess the code will not crash this way.
Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00



Le mar. 11 déc. 2018 à 21:46, James Croft <notifications@github.com> a
écrit :

**Comment by Mizux** [comment]
My bad I testes using the callback in line 180.
So, yes, as lperron suggested you must keep a ref on callback since ortools only keep a "pointer" on it in the C++ but the JVM/GC is not aware of this.

To reproduce it at 100% simply use
```java
model.setArcCostEvaluatorOfVehicle(manhattanCostCallback, vehicle);
System.gc(); // here the GC will reclaim manhatanCostCallback since model don't keep a "Java ref" on it
```

To fix use:
```java
 // Setting up vehicles
    final List<NodeEvaluator2> callbacks = new ArrayList();
    for (int vehicle = 0; vehicle < numberOfVehicles; ++vehicle) {
      final int costCoefficient = vehicleCostCoefficients.get(vehicle);
      NodeEvaluator2 manhattanCostCallback = new NodeEvaluator2() {
        @Override
        public long run(int firstIndex, int secondIndex) {
          try {
            Pair<Integer, Integer> firstLocation = locations.get(firstIndex);
            Pair<Integer, Integer> secondLocation = locations.get(secondIndex);
            return costCoefficient
                * (Math.abs(firstLocation.first - secondLocation.first)
                      + Math.abs(firstLocation.second - secondLocation.second));
          } catch (Throwable throwed) {
            logger.warning(throwed.getMessage());
            return 0;
          }
        }
      };
      model.setArcCostEvaluatorOfVehicle(manhattanCostCallback, vehicle);
      callbacks.add(manhattanCostCallback);
      System.gc(); // For test purpose
      model.cumulVar(model.start(vehicle), "time").setMin(vehicleStartTime.get(vehicle));
      model.cumulVar(model.end(vehicle), "time").setMax(vehicleEndTime.get(vehicle));
    }
    System.gc(); // For test purpose
```
